---
title: "Differential gene expression : microarray VS RNAseq"
date: "`r Sys.Date()`"
output:
    rmdformats::readthedown:
        number_sections: yes
        #toc: 3
        toc_float: yes
        theme: flatly
        highlight: zenburn
        df_print: paged
        code_folding: show
    rmdformats::html_clean:
        number_sections: yes
        #toc: 3
        toc_float: yes
        theme: flatly
        highlight: zenburn
        df_print: kable
        code_folding: show
    pdf_document:
        toc: yes
    html_document:
        number_sections: yes
        toc: yes
        toc_float: yes
        theme: flatly
        highlight: zenburn
        df_print: paged
        code_folding: show
---


```{r knitr_init, echo=FALSE, cache=FALSE}
library(knitr)
library(rmdformats)
library(DT)

## Global options
options(max.print="75")
opts_chunk$set(echo=FALSE,
	             cache=TRUE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE)
opts_knit$set(width=75)
```

# Context
## Why ?

In order to add a way to handle microarray data in Moomin, we first need to check if both differential gene expressions have similar distributions. Moomin uses a bayesian method to compute the LogFC : EBseq. The method that we are investigating is Cyber-t : its also a bayesian method, but for microarrays. If results between those two are significantly close, cyber-t method can be added in Moomin.

- [x] Cyber-t logFC
- [x] EBseq logFC
- [x] statistical tests (spearman $\rho$ and R-squared)
- [x] Interpretations
- [x] Integration in Moomin
- [ ] Flux analysis comparison in Moomin

## The dataset

We found a dataset from the paper ["*Escherichia coli* O157:H7 transcriptome datasets for comparison of RNA-seq and microarray platforms"](https://www.sciencedirect.com/science/article/pii/S2352340918315294?via%3Dihub) that has both RNAseq data and microarray data, for the same organism, with the same condition between samples. (With or without spinach)
A very-well organized git repo can be found at : https://github.com/TheMicroGirl/SakaiRNASeq/

# Differential expression analysis with microarray data and Cyber-t

```{r cyber-t dependencies}
library(limma)
library(gplots)
library(dplyr)
source("Microarray/cyber-t_R_scripts/bayesreg.R")
```

## Reading raw data

Raw data consist of 6 TSV files : 2 conditions and 3 replicates each.

Conditions are :

  * Leaf lysate
  * Minimal media
  
Limma functions are used here to extract relevant field.
```{r microarray input, warning = FALSE, echo=TRUE, include=FALSE}
setwd("Microarray/microarray_data/raw_data")
summaryTable<-read.delim("summaryTable.txt",check.names=FALSE,stringsAsFactors=FALSE)
microarrayData <-read.maimages(summaryTable[,"FileName"],source="agilent",green.only=TRUE)
```

## Probes filtering
Keeping only probes that are 10% brighter than negative probes.
```{r, echo=TRUE}
neg95 <- apply(microarrayData$E[microarrayData$genes$ControlType==-1,],2,function(x) quantile(x,p=0.95)) #95 percentile of negative probes
cutoff <- matrix(1.1*neg95,nrow(microarrayData),ncol(microarrayData),byrow=TRUE) # keep probes that are 10% brighter than the negative probes
filtered <- rowSums(microarrayData$E > cutoff) >= 3
noNegData <- microarrayData[microarrayData$genes$ControlType==0 & filtered,]
```


## Data processing
Rearranging data in a way cyber-t can use.
```{r microarray data processing, echo=TRUE}
df <- as.data.frame(noNegData$E)
df <- cbind(noNegData$genes$SystematicName,df)
df <- aggregate(df[,2:7], list(df$`noNegData$genes$SystematicName`), median) # Merging duplicates and keeping median values
colnames(df) <- c("gene_id", "C1", "C2", "C3", "E1", "E2", "E3" )
df <- df[c("gene_id", "E1", "E2", "E3", "C1", "C2", "C3")] # Rearranging df so our values are in the same order than RNAseq data
df <- format(df,  scientific = TRUE)
write.table(df, file="Microarray/microarray_data/processed_data/processed_MA_results.txt", col.names=FALSE, row.names=FALSE, sep =",", quote = FALSE)
```

## Call of Cyber-t "bayesT" function
Cyber-t is a method that compute a PPDE and LogFC for probes that got change in intensity (AKA differentially expressed genes).
Due to the large number of gene studied, we use 'winSize = 101' following cyber-t manual. 'bayes' parameter is set to '1' as we want to use a bayesian method.
```{r Call to cyber-t, echo=TRUE, results='hide'}
dataFile<-read.table("Microarray/microarray_data/processed_data/processed_MA_results.txt",sep=",", header = TRUE, row.names = 1)
cyber_t_results <- bayesT(dataFile,numC=3,numE=3,ppde=1, bayes=1, winSize=101, conf=4)
logFC <- log(cyber_t_results$meanE / cyber_t_results$meanC)
```

## Writing cyber-t results
Output genes matched with their logFC and PPDE.
```{r Rearranging cyber-t results and output, echo=TRUE}
cyber_t_output <- data.frame(rownames(cyber_t_results), cyber_t_results$ppde.p, cyber_t_results$fold)
cyber_t_output <- data.frame(rownames(cyber_t_results), cyber_t_results$ppde.p, logFC)
cyber_t_output <- cyber_t_output[order(cyber_t_output$cyber_t_results.ppde.p, decreasing=TRUE), ]
write.table(cyber_t_output, file="Microarray/results/cyber_t_results.txt", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

## Building reference table for later
Building a table that matches systematic gene names and probe names.
```{r}
# Re-using a table from older study to associate gene names with probe names. (gene names <-> probe names <-> ECs name)
blast <- read.table("misc/SakaiRNASeq/3.Microarray_Vs_RNASeq/3.table merge/final_blast_Table.txt", sep='\t', header=TRUE)
corresponding_names <- merge(dplyr::select(blast, query.name, subject), dplyr::select(noNegData$genes, ProbeName, SystematicName), by.x="query.name", by.y="ProbeName")
corresponding_names <- unique(corresponding_names)
write.table(corresponding_names, file="Comparison/reference_table.txt")
```

# Differential expression analysis with RNAseq data and EBseq
```{r dependecies, echo=FALSE}
library(memisc)
```
## Processing Raw data

Raw data can be found here : https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-7441/samples/

They were generated using these tools :

1. **Cutadapt** : Trimming adaptors 
2. **FastQC** : Quality check
3. **Bowtie2** : Indexing with reference genome
4. **FeatureCounts** : Counting with annotation

They consist of 6 pairs of fastq file (paired-end RNAseq) : 3 replicates per conditions.

```{r RNAseq plots and indexing genes and counts, warning=FALSE, output=FALSE, results='hide' }
par(mfrow=c(2,3))
files <- list.files("RNAseq/RNAseq_countings", pattern="*.tsv$", full.names=TRUE, recursive=FALSE)
get_title <- function(x) {
  title=cases(
    "sample 1 : Control"=grepl("ERR2955747", x),
    "sample 2 : Control"=grepl("ERR2955748", x),
    "sample 3 : Control"=grepl("ERR2955749", x),
    "sample 4 : with spinach leaves"=grepl("ERR2955750", x),
    "sample 5 : with spinach leaves"=grepl("ERR2955751", x),
    "sample 6 : with spinach leaves"=grepl("ERR2955752", x),
    "wtfisthat"=TRUE
  )
}


listplot <- function(x) {
  data = read.csv(file = x, header = TRUE, sep = "\t", skip=1)
  plot(data[,length(data)], type = "p",
       main = get_title(x),
       xlab = "genes",
       ylab = "counts"
  )
}
lapply(files,listplot)

getLastColumn <- function(x){
  content = read.csv(file = x, header = TRUE, sep = "\t", skip=1)
  return(content[,length(content)])
}
counts_list <- lapply(files,getLastColumn)

gene_ids <- read.csv(file = "RNAseq/RNAseq_countings/trimmed_ERR2955747.tsv", header = TRUE, sep = "\t", skip=1)[,1]
length(gene_ids)

EBseq_counts_df <- data.frame(gene_ids)
EBseq_counts_df <- cbind(EBseq_counts_df, counts_list)
colnames(EBseq_counts_df) <- c("gene_id", "ERR2955750", "ERR2955751", "ERR2955752", "ERR2955747", "ERR2955748", "ERR2955749")
EBseq_counts_df <- EBseq_counts_df[c("gene_id", "ERR2955747", "ERR2955748", "ERR2955749", "ERR2955750", "ERR2955751", "ERR2955752")]

write.table(EBseq_counts_df, file="RNAseq/processed_counts.txt", row.names=FALSE,sep ="\t", quote = FALSE )
```


## EBseq call : using 'EBTest' and 'GetDEResults' functions


```{r}
library(EBSeq)
counts = read.csv("RNAseq/processed_counts.txt", header = TRUE, sep = "\t")
data=as.matrix(counts)
rnames=counts[,1]
rownames(data)=rnames
data=data[,-1]
class(data)="numeric"

conditions = as.factor(c("C1","C1","C1","C2","C2","C2"))
Sizes=MedianNorm(data)
EBOut=EBTest(Data=data,Conditions=conditions,sizeFactors=Sizes, maxround=5)
EBDERes=GetDEResults(EBOut, FDR=0.05)

EBseq_results <-data.frame(EBDERes$PPMat, counts[2:7])
EBseq_results <- EBseq_results[order(EBseq_results$PPDE, decreasing=TRUE),]
```

## Checking EBseq results quality

### check convergence

```{r}
rounds = 5
diff(EBOut$Alpha[(rounds-2):rounds,])
diff(EBOut$Beta[(rounds-2):rounds,])
diff(EBOut$P[(rounds-2):rounds,])
```



### check model fitness
```{r}
par(mfrow=c(1,1))
QQP(EBOut)
DenNHist(EBOut)
```

## Writing results table
```{r}
EBFC <- PostFC(EBOut)
df_EBFC <- as.data.frame(EBFC["PostFC"])
df_EBFC <- data.frame(ECs=rownames(df_EBFC), logFC=log(df_EBFC[,1]))
df1 <- data.frame(EBDERes$PPMat)
df1$names <- rownames(EBDERes$PPMat)
final_EBseq_results <- merge(df1, df_EBFC, by.x="names", by.y="ECs")
write.table(final_EBseq_results, file="RNAseq/EBseq_results/final_EBseq_results.tsv", sep="\t", row.names = FALSE, quote = FALSE)
```

# Merging RNAseq and Microarray data
```{r, cleaning environment}
#rm(list=ls())
```

## Import data from our bayesian methods

### Import Cyber-t results for microarray

```{r}
cyber_t_results <- read.table("Microarray/results/cyber_t_results.txt")
colnames(cyber_t_results) <- c("gene_id", "cyber_t_PPDE", "cyber_t_logFC")
head(cyber_t_results)
cyber_t_results <- format(cyber_t_results, scientific = FALSE)
```

### Import EBseq results for RNAseq
```{r}
EBseq_results <- read.table("RNAseq/EBseq_results/final_EBseq_results.tsv", header=TRUE)
small_EBseq_results <- data.frame(gsub("ECs_", "ECs", EBseq_results$names), EBseq_results$PPDE, EBseq_results$logFC)
colnames(small_EBseq_results) <- c("gene_id", "EBseq_PPDE", "EBseq_logFC")

head(small_EBseq_results)
```
## Repair name discrepancies between RNAseq data and microarray data
```{r}
references <- read.table("Comparison/reference_table.txt")
EBseq_corresponding_genes <- merge(small_EBseq_results, references, by.x="gene_id", by.y="subject")
colnames(EBseq_corresponding_genes) <- c("ECs","EBseq_PPDE","EB_logFC","query.name", "SystematicName")
```

## Import data from the paper for comparison

### Import Limma results for microarray
```{r}
# Results from the paper for microarray (using limma)
limma_results<-read.table("misc/SakaiRNASeq/1.Microarray_Data_Analysis/microarrayTopTable.txt",sep='\t', header = TRUE)
# This file got an error in the naming of the columns. They are all misaligned of one col to the left. Need a fix
# bad_colnames <- colnames(limma_results)
# new_colnames <-c(bad_colnames[length(bad_colnames)], bad_colnames[1:length(bad_colnames)-1])
# colnames(limma_results) <- new_colnames

small_limma_results <- dplyr::select(limma_results, SystematicName, logFC, adj.P.Val)
colnames(small_limma_results) <- c("gene_id", "limma_logFC", "limma_Pval")
```

### Import EdgeR results for RNAseq
```{r}
rnaseq_results <- read.table("misc/SakaiRNASeq/2.RNASeq_Data_Analysis/3. DEG_analysis/RNAseqTopTable-Paired.txt", skip=1)
colnames(rnaseq_results) <- c("ECs", "logFC", "t", "P.Value", "adj.P.Val", "b")
EdgeR_corresponding_genes <- merge(rnaseq_results, references, by.x="ECs", by.y="subject")
EdgeR_corresponding_genes <- data.frame(EdgeR_corresponding_genes$SystematicName, EdgeR_corresponding_genes$P.Value,EdgeR_corresponding_genes$logFC)
colnames(EdgeR_corresponding_genes) <- c("gene_id", "EdgeR_Pval", "EdgeR_logFC")
```
## Merging all results in one big df
```{r}
summary <- merge(cyber_t_results, EBseq_corresponding_genes, by.x="gene_id", by.y="SystematicName")
summary <- merge(summary, EdgeR_corresponding_genes, by="gene_id") 
summary <- merge(summary, small_limma_results, by="gene_id")
summary <- summary[c("gene_id", "cyber_t_logFC", "EB_logFC", "limma_logFC", "EdgeR_logFC", "cyber_t_PPDE", "EBseq_PPDE", "limma_Pval", "EdgeR_Pval")]
```
## Filtering out unsignificant values
```{r}
dim(summary)
only_significant_values<-summary[(summary$cyber_t_PPDE > 0.95 & summary$EBseq_PPDE > 0.95 & summary$limma_Pval < 0.05 & summary$EdgeR_Pval < 0.05),]
dim(only_significant_values)
only_significant_values[complete.cases(only_significant_values), ]
only_significant_values <- na.omit(only_significant_values)
dim(only_significant_values)
```
## Filtering out missing values
```{r}
dim(summary)
summary[complete.cases(summary), ]
summary <- na.omit(summary)
dim(summary)
```

# Results comparison

## Limma Vs EdgeR
```{r}
plot(summary$limma_logFC, summary$EdgeR_logFC,
     main = "Limma vs EdgeR logFC comparison",
     xlab = "Limma logFC",
     ylab = "EdgeR logFC",
     xlim = c(-6,6),
     ylim = c(-6,6))
rho_limma_vs_FC <- cor.test(~ as.numeric(summary$limma_logFC) + as.numeric(summary$EdgeR_logFC), data=summary, method = "spearman", continuity = TRUE)$estimate
r_squared_limma_vs_FC <- summary(lm(summary$limma_logFC ~ summary$EdgeR_logFC, data=summary))$adj.r.squared
rho_limma_vs_FC2 <- cor.test(~ as.numeric(only_significant_values$limma_logFC) + as.numeric(only_significant_values$EdgeR_logFC), data=only_significant_values, method = "spearman", continuity = TRUE)$estimate
r_squared_limma_vs_FC2 <- summary(lm(only_significant_values$limma_logFC ~ only_significant_values$EdgeR_logFC, data=only_significant_values))$adj.r.squared

```

## EBseq Vs EdgeR
```{r}

length(summary$EdgeR_logFC)
length(summary$EB_logFC)
length(as.numeric(summary$EB_logFC))
length(as.numeric(summary$EdgeR_logFC))
plot(summary$EB_logFC, summary$EdgeR_logFC, 
     main = "EBseq vs EdgeR logFC comparison", 
     xlab = "EBseq logFC",
     ylab = "EdgeR logFC",
     xlim = c(-6,6),
     ylim = c(-6,6))
rho_EB_vs_FC <- cor.test(~ as.numeric(summary$EB_logFC) + as.numeric(summary$EdgeR_logFC), data=summary, method = "spearman", continuity = TRUE)$estimate
r_squared_EB_vs_FC <- summary(lm(summary$EB_logFC ~ summary$EdgeR_logFC, data=summary))$adj.r.squared
```

## Cyber-t Vs Limma
```{r}
plot(summary$cyber_t_logFC, summary$limma_logFC, 
     main = "Cyber-t vs limma logFC comparison", 
     xlab = "Cyber-t logFC",
     ylab = "limma logFC",
     xlim = c(-6,6),
     ylim = c(-6,6))
rho_cyber_t_vs_limma <- cor.test(~ as.numeric(summary$cyber_t_logFC) + as.numeric(summary$limma_logFC), data=summary, method = "spearman", continuity = TRUE)$estimate
r_squared_cyber_t_vs_limma <- summary(lm(summary$cyber_t_logFC ~ summary$limma_logFC, data=summary))$adj.r.squared
```

## Cyber-t Vs EBseq
```{r}
plot(summary$cyber_t_logFC, summary$EB_logFC, 
     main = "Cyber-t vs EBseq logFC comparison", 
     xlab = "Cyber-t logFC",
     ylab = "EBseq logFC",
     xlim = c(-6,6),
     ylim = c(-6,6))
rho_cyber_t_vs_EB <- cor.test(~ as.numeric(summary$cyber_t_logFC) + as.numeric(summary$EB_logFC), data=summary, method = "spearman", continuity = TRUE)$estimate
r_squared_cyber_t_vs_EB <- summary(lm(summary$cyber_t_logFC ~ summary$EB_logFC, data=summary))$adj.r.squared
rho_cyber_t_vs_EB2 <- cor.test(~ as.numeric(only_significant_values$cyber_t_logFC) + as.numeric(only_significant_values$EB_logFC), data=only_significant_values, method = "spearman", continuity = TRUE)$estimate
r_squared_cyber_t_vs_EB2 <- summary(lm(only_significant_values$cyber_t_logFC ~ only_significant_values$EB_logFC, data=only_significant_values))$adj.r.squared
```

## Table of results
```{r echo = FALSE, results = 'asis'}
rho <- c(rho_limma_vs_FC,rho_EB_vs_FC,rho_cyber_t_vs_limma,rho_cyber_t_vs_EB,rho_limma_vs_FC2,rho_cyber_t_vs_EB2)
r_squared <- c(r_squared_limma_vs_FC,r_squared_EB_vs_FC,r_squared_cyber_t_vs_limma,r_squared_cyber_t_vs_EB,r_squared_limma_vs_FC2,r_squared_cyber_t_vs_EB2)
stat_table <- tibble(rho, r_squared)
colnames(stat_table) <- c("Spearman : rho", "R²")
rownames(stat_table) <- c("Limma / EdgeR","EBseq / EdgeR","Cyber-t / Limma","Cyber-T / EBseq", "limma / FC (above threshold)", "Cyber-t / EBseq (above threshold)")
kable(stat_table)
```



## Volcano plots
```{r, echo=TRUE}
with(summary, plot(cyber_t_logFC, cyber_t_PPDE, pch=20, main="Cyber-t", xlim=c(-2,2)))
with(summary, plot(limma_logFC, limma_Pval, pch=20, main="Limma",  xlim=c(-2,2)))
with(summary, plot(EB_logFC, EBseq_PPDE, pch=20, main="EBseq",  xlim=c(-2,2)))
with(summary, plot(EdgeR_logFC, EdgeR_Pval, pch=20, main="EdgeR",  xlim=c(-2,2)))
```

# Format and export data for Moomin input
```{r}
EB_exp_mat_for_moomin <- data.frame(only_significant_values$gene_id, only_significant_values$EBseq_PPDE, only_significant_values$EB_logFC)
colnames(EB_exp_mat_for_moomin) <- c("GeneID", "PPDE", "FC")
write.table(EB_exp_mat_for_moomin, file = "Moomin_input/EBseq_res.tsv", quote=F, sep="\t", row.names = F)
```
```{r}
cyber_t_exp_mat <- data.frame(only_significant_values$gene_id, only_significant_values$cyber_t_PPDE, only_significant_values$cyber_t_logFC)
colnames(cyber_t_exp_mat) <- c("GeneID", "PPDE", "FC")
write.table(cyber_t_exp_mat, file = "Moomin_input/cyber_t_res.tsv", quote=F, sep="\t", row.names = F)
```

# Compare Moomin results from EBSeq or Cyber-T 
```{r}
library("rjson")
RNASeq_moomin_output <- t(as.data.frame(fromJSON(file = "../Cyber-t_integration_to_Moomin/output/EBseq_output_stoich.json")))
RNA_output <- cbind.data.frame(rownames(RNASeq_moomin_output), RNASeq_moomin_output)

microarray_moomin_output <- t(as.data.frame(fromJSON(file = "../Cyber-t_integration_to_Moomin/output/cyber_t_output_stoich.json")))
microarray_output <- cbind.data.frame(rownames(microarray_moomin_output), microarray_moomin_output)

merged_df <- merge(RNASeq_moomin_output, microarray_moomin_output, by="row.names", all.x=TRUE)
colnames(merged_df) <- c("reaction_ID", "RNA_color", "microarray_color")

# Number of reaction with same color ?
length(which(merged_df$RNA_color == merged_df$microarray_color))
no_good <- merged_df[merged_df$RNA_color != merged_df$microarray_color,]
erf <- no_good[no_good$microarray_color == 0,]
dim(no_good)
dim(erf)
rho_dabb <- cor.test(~ as.numeric(merged_df$RNA_color) + as.numeric(merged_df$microarray_color), data=merged_df, method = "spearman", continuity = TRUE)$estimate
rho_dabb

plot(merged_df$RNA_color, merged_df$microarray_color, 
     main = "Limma vs EdgeR logFC comparison", 
     xlab = "Limma logFC",
     ylab = "EdgeR logFC")
```

